torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3004,6 +3005,164 @@ struct to_ir {
+          // constructor with a dict literal. It makes much more sense
+          // to transform the dict literal into a list of tuples so that
+          // we can use the existing constructors
+          if (!apply.inputs().empty() &&
+              apply.inputs()[0].kind() == TK_DICT_LITERAL) {
+            auto dict_lit = DictLiteral(apply.inputs()[0]);
+            std::vector<Expr> zipped;
+            for (auto key_it = dict_lit.key_inputs().begin(),
+                      val_it = dict_lit.value_inputs().begin();
+                 key_it != dict_lit.key_inputs().end();
+                 ++key_it, ++val_it) {
+              auto tuple_inputs =
+                  List<Expr>::create(apply.range(), {*key_it, *val_it});
+              auto tuple = TupleLiteral::create(apply.range(), tuple_inputs);
+              zipped.push_back(tuple);
+            }
+            auto ll_values = List<Expr>::create(apply.range(), zipped);
+            auto ll = ListLiteral::create(apply.range(), ll_values);
+            auto expr_list = List<Expr>::create(apply.range(), {ll});
+            // Change `apply` to a new Apply node holding a list of
+            // tuples
+            apply = Apply::create(
+                apply.range(), apply.callee(), expr_list, apply.attributes());
+          }
+
+          // If we have kwargs to include, we'll take a similar approach
+          // to the above logic and standardize the Apply node
+          if (!apply.attributes().empty()) {
+            std::vector<Expr> exprs;
+            // Gather all the existing tuples in the input iterable
+            if (!apply.inputs().empty()) {
+              auto tuple_list = ListLiteral(apply.inputs()[0]).inputs();
+              for (auto tuple : tuple_list) {
+                exprs.push_back(tuple);
+              }
+            }
+            // Create tuples out of each kwarg and gather them as well
+            for (auto attr : apply.attributes()) {
+              auto k = StringLiteral::create(apply.range(), attr.name().name());
+              auto v = attr.value();
+              auto tuple_inputs = List<Expr>::create(apply.range(), {k, v});
+              auto tuple = TupleLiteral::create(apply.range(), tuple_inputs);
+              exprs.push_back(tuple);
+            }
+            auto expr_list = List<Expr>::create(apply.range(), {exprs});
+            auto ll = ListLiteral::create(apply.range(), expr_list);
+            auto new_inputs = List<Expr>::create(apply.range(), {ll});
+            auto new_kwargs = List<Attribute>::create(apply.range(), {});
+            apply = Apply::create(
+                apply.range(), apply.callee(), new_inputs, new_kwargs);
+          }
+
+          checkApplyNumInputs(apply, 1);
+
+          auto iter_input = emitSugaredExpr(apply.inputs()[0], 1);
+
+          // If possible, dispatch to the `aten::dict` builtin op to
+          // avoid perf slowdown on existing uses
+          if (auto simple = asSimple(iter_input)) {
```

Comment:
I don't know if this works - `dict((1, 2), (3, 4))`. Here the input is simple but we can't dispatch to aten::dict because the  input isn't a list. The dispatch to aten::list was to avoid slowdown and because some slow test took  forever with the new change.  But since we're introducing `dict()` from start - we have a little more leeway to not do *the most optimized* implementation from start.  It  might  be better to keep the PR simpler at first and then follow up and optimize as needed.

Reply:
I don't think `dict((1, 2), (3, 4))` is valid Python--I just tried it in the interpreter.